libimage/filters.go

```diff
@@ -265,62 +296,90 @@ func filterManifest(ctx context.Context, value bool) filterFunc {
+	return func(img *Image) (bool, []string, error) {
 		// Empty reference filters, return true
 		if len(wantedReferenceMatches) == 0 && len(unwantedReferenceMatches) == 0 {
-			return true, nil
+			return true, nil, nil
 		}
 
-		unwantedMatched := false
 		// Go through the unwanted matches first
+		// TODO 6.0 podman: remove unwanted matches from the output names. https://github.com/containers/common/pull/2413#discussion_r2031749013
 		for _, value := range unwantedReferenceMatches {
-			matches, err := imageMatchesReferenceFilter(r, img, value)
+			names, err := getMatchedImageNames(r, img, value)
 			if err != nil {
-				return false, err
+				return false, nil, err
 			}
-			if matches {
-				unwantedMatched = true
+			if len(names) > 0 {
+				return false, nil, nil
 			}
 		}
 
+		namesThatMatch := slices.Clone(img.Names())
 		// If there are no wanted match filters, then return false for the image
 		// that matched the unwanted value otherwise return true
 		if len(wantedReferenceMatches) == 0 {
-			return !unwantedMatched, nil
+			return true, namesThatMatch, nil
 		}
 
 		// Go through the wanted matches
 		// If an image matches the wanted filter but it also matches the unwanted
 		// filter, don't add it to the output
+		matchedNames := map[string]struct{}{}
+
 		for _, value := range wantedReferenceMatches {
-			matches, err := imageMatchesReferenceFilter(r, img, value)
+			names, err := getMatchedImageNames(r, img, value)
 			if err != nil {
-				return false, err
+				return false, nil, err
 			}
-			if matches && !unwantedMatched {
-				return true, nil
+
+			for name := range names {
+				matchedNames[name] = struct{}{}
 			}
 		}
+		if len(matchedNames) > 0 {
+			// Removes non-compliant names from image names
+			namesThatMatch = slices.DeleteFunc(namesThatMatch, func(name string) bool {
+				_, ok := matchedNames[name]
+				return !ok
+			})
+			return true, namesThatMatch, nil
+		}
 
-		return false, nil
+		return false, nil, nil
 	}
 }
 
-// imageMatchesReferenceFilter returns true if an image matches the filter value given
-func imageMatchesReferenceFilter(r *Runtime, img *Image, value string) (bool, error) {
-	lookedUp, _, _ := r.LookupImage(value, nil)
+// getMatchedImageNames returns a list of matching image names that match the specified filter value, or an empty list if the image does not match the filter.
+func getMatchedImageNames(r *Runtime, img *Image, value string) (map[string]struct{}, error) {
+	lookedUp, resolvedName, _ := r.LookupImage(value, nil)
 	if lookedUp != nil {
 		if lookedUp.ID() == img.ID() {
-			return true, nil
+			out := map[string]struct{}{}
+
+			imageName, _, digestFound := strings.Cut(resolvedName, "@")
+			if digestFound {
+				// If resolvedName contains a digest, return all names that match imageName.
+				// Note: Tags are ignored. (e.g. busybox:ignoreme@sha256:1234567890...)
+				for _, fullName := range lookedUp.Names() {
+					name, _, _ := strings.Cut(fullName, ":")
```

Comment:
_Don’t parse names using custom string logic_ if at all possible: This will cut `quay.io/foo@sha256:…` after `sha256`.

In this case that _seems_ to be harmless because `imageName` will never contain a @, but that’s hard to follow.

Below, the code relies on `NamesReferences` and uses those parsed values, I think that makes much more sense.